operator/builtin/input/file/design.md

```diff
@@ -0,0 +1,146 @@
+
+In some rare circumstances, a logger may print a very verbose preamble to each log file. When this occurs, fingerprinting may fail to differentiate files from one another. This can be overcome by customizing the size of the fingerprint using the `fingerprint_size` setting.
+
+# Readers
+
+Readers are a convenience struct, which exist for the purpose of managing files and their associated metadata. 
+
+### Contents
+
+A Reader contains the following:
+- File handle (may be open or closed)
+- File fingerprint
+- File offset (aka checkpoint)
+- File path
+- Decoder (dedicated instance to avoid concurrency issues)
+
+### Functionality
+
+As implied by the name, Readers are responsible for consuming data as it is written to a file.
+
+Before a Reader begins consuming, it will seek the file's last known offset. Then it will begine reading from there.
+
+While a file is shorter than the length of a fingerprint, its Reader will continuously append to the fingerprint, as it consumes newly written data.
+
+A Reader consumes a file using an `bufio.Scanner`, with the Scanner's buffer size defined by the `max_log_size`setting, and the Scanner's split func defined by the `multiline` setting. 
+
+As each log is read from the file, it is decoded according to the `encoding` function, and then emitted from the operator. 
+
+The Reader's offset is updated accordingly when a log is emitted.
+
+
+### Persistence
+
+Readers are always instantiated with an open file handle. Eventually, the file handle is closed, but the Reader is not discarded. Rather, it is maintained for a fixed number of "poll cycles" (see Polling section below) as a reference to the file's metadata, which may be useful for detecting files that have been moved or copied, and for recalling metadata such as the file's previous path.
+
+Readers are maintained for a fixed period of time, and then discarded.
+
+When the `file_input` operator makes use of a persistence mechanism to save and recall its state, it is simply Setting and Getting a slice of Readers. These Readers contain all the information necessary to pick up exactly where the operator left off.
+
+
+# Polling
+
+The file system is polled on a regular interval, defined by the `poll_interval` setting. 
+
+Each poll cycle runs through a series of steps which are presented below.
+
+At a very high level, each poll cycle operates as three phases:
+1. Finish work that was started in the previous poll cycle.
+2. Begin work that will carry over to the next cycle.
+3. Allow some time to pass.
+
+Because state is carried over from one poll cycle to the next, the following detailed description is presented starting mid-cycle, at a point where there is no preexisting state.
+
+
+### Detailed Poll Cycle
+
+_Note: The following steps are presented starting mid poll cycle._
```

Comment:
I am a bit confused: why isn't the description bellow start from "Start Poll Cycle" step? I understand it is cyclical but it seems a bit more natural than starting mid cycle?

Reply:
Point taken. I'll reorganize it to just describe the cycle from start to finish.

I thought it might be more confusing to start the description with steps that are dependent on the state of the previous cycle, since understanding these steps requires understanding the rest of the poll cycle. 

Later, I'll explore whether these first few steps can just be moved to the end of the cycle, which would be a lot easier to reason about anyways.